js/style/style_layer.js

```diff
@@ -180,56 +180,53 @@ StyleLayer.prototype = util.inherit(Evented, {
         return false;
     },
 
-    // update classes
-    cascade: function(classes, options, globalTransitionOptions, animationLoop) {
-        var oldTransitions = this._paintTransitions;
-        var newTransitions = this._paintTransitions = {};
-        var that = this;
-
-        // Apply new declarations in all active classes
-        for (var klass in this._paintDeclarations) {
-            if (klass !== "" && !classes[klass]) continue;
-            for (var name in this._paintDeclarations[klass]) {
-                applyDeclaration(name, this._paintDeclarations[klass][name]);
-            }
+    // update all paint transitions and layout/paint constant values
+    cascade: function(classes, options, globalOptions, animationLoop) {
+        var declarations = util.extend({}, this._paintDeclarations['']);
+        for (var klass in classes) {
```

Comment:
What guarantees do we make about the order that style classes are applied to the map? We should document this explicitly somewhere.